GateSeal v2.0

Alexvozhak opened the pull request:
## 🔧 Overview

This PR finalizes the implementation of **GateSeal v2** mechanics as described in the [ADR](https://www.notion.so/ADR-X-GateSeal-v2-216bf633d0c9809994ebd484c6334e42).

---
### ✅ Business Logic
- All business requirements from the ADR are fully implemented.
- All contract changes are isolated in dedicated `V2` versions

📎 [See ADR block here](https://www.notion.so/ADR-X-GateSeal-v2-216bf633d0c9809994ebd484c6334e42?source=copy_link#240bf633d0c98051b1adf523732e458f)

---
### 📘 Docs
- Updated [README.md](https://github.com/lidofinance/gate-seals/pull/34/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5)
  - Redrawn the dependency diagram.
  - Removed typos.
  - All business-related changes are also reflected in the README.
- Added [AGENTS.md](https://github.com/lidofinance/gate-seals/pull/36/files#diff-a54ff182c7e8acf56acfd6e4b9c3ff41e2c41a31c9b211b2deb9df75d9a478f9) for the Codex agent

---
### 🧱 Dependency & Toolchain Updates
- Vyper bumped from `0.3.7` → `0.4.2`
 👉 See [release notes](https://docs.vyperlang.org/en/stable/release-notes.html)
- Ape plugins upgraded from `0.6.x` → `0.8.x`  
  - EVM hardfork updated from **Shanghai** → **Cancun**  
  👉 See [proof in Vyper release notes](https://github.com/vyperlang/vyper/blob/6ecdb3c01088ebd0060ccdc65a5a4c231e6340cc/docs/release-notes.rst?plain=1#L210)
- Hardhat bumped from `2.12.6` → `2.24.1`  
  - Cancun support introduced in `v2.20.0`  
  👉 See [Hardhat changelog №1](https://github.com/NomicFoundation/hardhat/blob/2bbb705f810ac4579597ccaf51bc5f20bcd0d471/packages/hardhat-core/CHANGELOG.md?plain=1#L213)  
  - Minimum required Node.js version raised from `16` → `18`. Updated to `22 LTS`  
  👉 See [Hardhat changelog №2](https://github.com/NomicFoundation/hardhat/blob/2bbb705f810ac4579597ccaf51bc5f20bcd0d471/packages/hardhat-core/CHANGELOG.md?plain=1#L196)
- **Husky** pre-commit hooks configured with **commitlint** to enforce commit message conventions  
  👉 [see changes](https://github.com/lidofinance/gate-seals/pull/36/files#diff-292bac03df29207ee254ba91cfc30951b10c50c0a9fff3a0768778cebe5f4c4a)



### 🐛 Bugfix
 - Updated the tests to always use the current chain.pending_timestamp, avoiding reliance on static fixture values.
 - Fixed compatibility issue caused by non-deterministic config handling in `ape-hardhat@^0.6.4`.  
Starting from `v0.6.5`, config behavior was made deterministic:
👉 See [ape-hardhat@v0.6.5 release notes](https://github.com/ApeWorX/ape-hardhat/releases/tag/v0.6.5)

Alexvozhak committed:
chore: Update dependencies and apply necessary project adjustments

Alexvozhak committed:
chore: Update poetry.lock and yarn.lock

Alexvozhak committed:
refactor: fix typo EIP 522 to EIP 5202

Alexvozhak committed:
chore: Bump Python to 3.10.13 and update poetry.lock

Alexvozhak committed:
fix: remove duplicate constant definition

Alexvozhak committed:
chore: remove .python-version file

Alexvozhak committed:
chore: update Node.js to v22 LTS

Alexvozhak committed:
docs: fix typo in README.md

Alexvozhak committed:
feat: implement GateSeal v2 business logic

Alexvozhak committed:
chore: update dependencies in yarn.lock

Alexvozhak committed:
docs: add Codex Agent instructions for setup, testing, and deployment

Alexvozhak committed:
chore: rename 'extension' to 'prolongation'

Alexvozhak committed:
fix: remarks after the review
- restored committee's right to prolong
- changed terminology from extension to prolongation

Alexvozhak committed:
feat: add maximum prolongations limit to GateSeal contract

Alexvozhak committed:
feat: add maximum prolongation duration limit

Alexvozhak committed:
fix: rewrite tests for strictness
-  fix systemic timestamp handling issue also

Alexvozhak committed:
docs: update README.md
 - clarify dependency diagram

Alexvozhak committed:
Merge branch 'v2_dependencies' into feature/v2

Alexvozhak committed:
Merge branch 'v2_business' into feature/v2

Alexvozhak committed:
docs: fix typos in README.md

failingtwice commented on contracts/GateSeal.vy:
```diff
@@ -55,20 +60,28 @@ MAX_SEAL_DURATION_SECONDS: constant(uint256) = SECONDS_PER_DAY * MAX_SEAL_DURATI
+# Maximum number of prolongations allowed
+MAX_PROLONGATIONS: constant(uint256) = 5
+
+# Maximum duration of a single prolongation is 6 months
+MAX_PROLONGATION_DURATION_DAYS: constant(uint256) = 180
+MAX_PROLONGATION_DURATION_SECONDS: constant(uint256) = SECONDS_PER_DAY * MAX_PROLONGATION_DURATION_DAYS
```
Why this number? Why can't the committee prolong the expiry for another year, for example? 🤔 
If they prolonged the gateseal, they've proven that the committee is in tact. Perhaps then we can prolong for the same duration as the original expiry?

Alexvozhak replied:
I believe that over time, the risk increases that the committee may lose its effectiveness and viability. Therefore, as its initial term is about to expire, it should be required to demonstrate its continued value to the DAO more frequently.

@iamnp agrees with me on this point.

failingtwice commented on contracts/GateSeal.vy:
```diff
@@ -167,13 +220,19 @@ def seal(_sealables: DynArray[address, MAX_SEALABLES]):
+                sealed_for=SEAL_DURATION_SECONDS,
+                sealable=sealable,
+                sealed_at=block.timestamp
+            )
```
Probably we can rethink what we log here: probably redundant to log gate seal address and block timestamp

Alexvozhak replied:
[done it ](https://github.com/lidofinance/gate-seals/pull/36/commits/d302d3d237977420e58ebb916005b7d11b8a42b8)

failingtwice commented on contracts/GateSeal.vy:
```diff
@@ -167,13 +220,19 @@ def seal(_sealables: DynArray[address, MAX_SEALABLES]):
+            abi_encode(SEAL_DURATION_SECONDS, method_id=method_id("pauseFor(uint256)")),
             max_outsize=32,
             revert_on_failure=False
         )
```
i think the above comment might be outdated since they have fixed this bug.

failingtwice commented on contracts/GateSeal.vy:
```diff
@@ -167,13 +220,19 @@ def seal(_sealables: DynArray[address, MAX_SEALABLES]):
             sealable,
-            _abi_encode(SEAL_DURATION_SECONDS, method_id=method_id("pauseFor(uint256)")),
+            abi_encode(SEAL_DURATION_SECONDS, method_id=method_id("pauseFor(uint256)")),
             max_outsize=32,
```
`max_outsize=32`

this was only needed to bypass the bug. Now that it's fixed and we don't actually need the return data, we can set it to `0`.

failingtwice commented on contracts/GateSeal.vy:
```diff
@@ -191,6 +250,7 @@ def _is_expired() -> bool:
 @internal
 def _expire_immediately():
     self.expiry_timestamp = block.timestamp
+    self.prolongations_remaining = 0
```
Might be redundant since expired gate seal cannot be prolonged

failingtwice commented on contracts/GateSeal.vy:
```diff
@@ -155,7 +208,7 @@ def seal(_sealables: DynArray[address, MAX_SEALABLES]):
     failed_indexes: DynArray[uint256, MAX_SEALABLES] = []
```
Still no dynamic length strings in Vyper? If not, maybe a bitmap is less awkward to use

Alexvozhak replied:
[done it](https://github.com/lidofinance/gate-seals/pull/36/commits/d302d3d237977420e58ebb916005b7d11b8a42b8)

Alexvozhak committed:
fix: simplify raw_call usage

Alexvozhak committed:
fix: remove redundant prolongations_remaining reset

Alexvozhak committed:
fix: streamline raw_call implementation

Alexvozhak committed:
Update lifetime settings and tests

Alexvozhak committed:
Update GateSeal durations and constructor

Alexvozhak committed:
chore: reduce diff

TheDZhon commented on contracts/GateSeal.vy:
I'd renamed the contract to `GateSealV2.py` and the corresponding factory too, limiting the surface of possible operational errors when it's needed to distinguish the gate seals deployed rapidly.

Alexvozhak replied:
[done it](https://github.com/lidofinance/gate-seals/pull/36/commits/08c7b0e1f2b14b99c2edc65fd85c2d0ccda9736f)

Alexvozhak committed:
Document legacy contract exclusion

Alexvozhak committed:
Document legacy contract exclusion

Alexvozhak committed:
refactor: use bitmap for failed indexes in error messages

Alexvozhak committed:
Merge branch 'codex/refactor-contracts-for-gatesealv2-deployment' into feature/v2

Alexvozhak committed:
chore: add husky commitlint

Alexvozhak committed:
feat: update GateSeal V2 parameters

Alexvozhak committed:
feat: aligned with the latest ADR

Alexvozhak committed:
feat: aligned ADR v2

Alexvozhak committed:
chore: bump Vyper from 0.4.1 to 0.4.2

Alexvozhak committed:
fix: remove redundant fields from Sealed event

failingtwice commented on contracts/GateSealV2.vy:
```diff
@@ -0,0 +1,324 @@
+
+"""
+@title GateSealV2
+@author mymphe
```
outdated ;)

Alexvozhak replied:
[done](https://github.com/lidofinance/gate-seals/pull/36/commits/1679d9994175f3d486327d2562704e747d48baa3)

failingtwice commented on contracts/GateSealV2.vy:
```diff
@@ -0,0 +1,324 @@
+     the contract(s) immediately for a set duration, e.g. one week, which gives
+     the DAO the time to analyze the situation, decide on the course of action,
+     hold a vote, implement fixes, etc. GateSeal can only be used once.
+     GateSeal assumes that they have the permission to pause the contracts.
```
Prolly we can get rid of the V1 comment and write a new one for V2 from scratch

Alexvozhak replied:
[done](https://github.com/lidofinance/gate-seals/pull/36/commits/1a4fad75c76117dfa22c007b08a8f0faac63894f)

failingtwice commented on contracts/GateSealV2.vy:
```diff
@@ -0,0 +1,324 @@
+event Sealed:
+    sealed_by: address
+    sealed_for: uint256
+    sealable: address
```
Probably better to put the sealable first

Alexvozhak replied:
https://github.com/lidofinance/gate-seals/pull/36#discussion_r2242655893 - i also prefer to remove it

failingtwice commented on contracts/GateSealV2.vy:
```diff
@@ -0,0 +1,324 @@
+    assert len(_sealables) > 0, "sealables: empty list"
+    assert _initial_lifetime_seconds >= MIN_INITIAL_LIFETIME_SECONDS, "initial lifetime: too short"
+    assert _initial_lifetime_seconds <= MAX_INITIAL_LIFETIME_SECONDS, "initial lifetime: exceeds max"
+    assert _prolongations >= 0, "max prolongations: must be non-negative"
```
Redundant check: it's an unsigned int, it can't be negative

Alexvozhak replied:
[done](https://github.com/lidofinance/gate-seals/pull/36/commits/7c3f2cd8fc602529ee5e3fa0f3047891e9f48158)

failingtwice commented on contracts/GateSealV2.vy:
```diff
@@ -0,0 +1,324 @@
+    assert not self._is_expired(), "gate seal: expired"
+    assert self.prolongations_remaining > 0, "prolongations: exhausted"
+    start: uint256 = self.expiry_timestamp - DAO_RESERVE_SECONDS - PROLONGATION_WINDOW_SECONDS
+    end: uint256 = self.expiry_timestamp - DAO_RESERVE_SECONDS
```
We can use the view functions here

Alexvozhak replied:
[done](https://github.com/lidofinance/gate-seals/pull/36/commits/ab3b23d22346edd7274c98df162be9e3868d5ffc)

failingtwice commented on contracts/GateSealV2.vy:
```diff
@@ -0,0 +1,324 @@
+# The time window before expiry during which prolongation is allowed.
+# The initial lifetime in seconds, set at deployment to align the first
+# expiration with the chosen prolongation window.
+INITIAL_LIFETIME_SECONDS: immutable(uint256)
```
Why do we need this variable?

Alexvozhak replied:
It helps keep all GateSeals in sync — even if deployed at different times, they expire together.  
Easier for the committee to handle.  
(See [diagram](https://www.notion.so/ADR-X-GateSeal-v2-216bf633d0c9809994ebd484c6334e42#23ebf633d0c980a7b0a0de8ac3c56266) for context.)

failingtwice commented on contracts/GateSealFactoryV2.vy:
```diff
@@ -0,0 +1,79 @@
+
+"""
+@title GateSealFactoryV2
+@author mymphe
```
not me :)

Alexvozhak replied:
[done](https://github.com/lidofinance/gate-seals/pull/36/commits/1679d9994175f3d486327d2562704e747d48baa3)

failingtwice commented on contracts/GateSealFactoryV2.vy:
```diff
@@ -0,0 +1,79 @@
+     functioning GateSeal itself but only its initcode.
+
+     More on blueprints
+     https://docs.vyperlang.org/en/v0.4.1/built-in-functions.html#chain-interaction
```
wrong vyper version in the link

Alexvozhak replied:
[done](https://github.com/lidofinance/gate-seals/pull/36/commits/258f9838c346ac56ea8e3845d8eb966a713ff402)

failingtwice commented on contracts/GateSealV2.vy:
```diff
@@ -0,0 +1,324 @@
+            revert_on_failure=False
+        )
+        
+        if success and staticcall IPausableUntil(sealable).isPaused():
+            log Sealed(
+                sealed_by=SEALING_COMMITTEE,
+                sealed_for=SEAL_DURATION_SECONDS,
+                sealable=sealable,
+            )
+        else:
+            failed_indexes.append(sealable_index)
```
Since now we don't specify the sealables, perhaps it doesn't make sense to fire individual events

Alexvozhak replied:
[done](https://github.com/lidofinance/gate-seals/pull/36/commits/43a9baa0f1dcbf495b0e7f716032035122633a00)

TheDZhon replied:
???

Alexvozhak replied:
As we’ve decided to keep both implementations (seal → seal_some and seal_all), I’d prefer to leave it as is

Alexvozhak committed:
fix: remove sealable field from Sealed event

Alexvozhak committed:
fix: update Vyper documentation link to version 0.4.2

Alexvozhak committed:
chore: update author information in GateSeal contracts

Alexvozhak committed:
refactor: dry prolongation window calculations

Alexvozhak committed:
refactor: remove redundant check for non-negative prolongations

Alexvozhak committed:
chore: improve documentation clarity in GateSealV2 contract

Alexvozhak committed:
feat: aligned ADR v3

- implement configurable prolongation windows with full deployment flexibility
- Revert the sealable address in the Sealed event for backward compatibility in alert monitoring
- enhance test coverage
- update documentation

failingtwice commented on contracts/GateSealV2.vy:
```diff
@@ -0,0 +1,357 @@
+        revert_on_failure=False
+    )
+    
+    assert success, "sealable: does not implement IPausableUntil interface"
```
The presence of `isPaused()` on the contract does not guarantee that it implements `pauseFor(uint256)`

failingtwice commented on contracts/GateSealV2.vy:
```diff
@@ -0,0 +1,357 @@
+    
+    for sealable: address in _sealables:
+        assert sealable != empty(address), "sealables: includes zero address"
+        self._validate_sealable_interface(sealable)
```
I don't think this validation brings a lot. Even if it did check for the presence of `pauseFor(uint256)` interface (which it doesn't, see comment for _validate_sealable_interface), it does not guarantee that the sealable will have this function at the moment of sealing.

Alexvozhak replied:
[ok, reverted](https://github.com/lidofinance/gate-seals/pull/36/commits/76aacc2e343c5efd5385e84d3f45bfacb49fa3c9)

failingtwice commented on contracts/GateSealV2.vy:
```diff
@@ -0,0 +1,357 @@
+    start: uint256 = self._get_prolongation_window_start()
+    end: uint256 = self._get_prolongation_window_end()
+    assert block.timestamp >= start, "prolongation window: too early"
+    assert block.timestamp <= end, "prolongation window: expired"
```
Prolongation window end timestamp is inclusive, while gateseal expiry timestamp is exclusive.

block.timestamp >= prolong_window_end_timestamp = still valid
block.timestamp >= expiry_timestamp = expired

Alexvozhak replied:
[done](https://github.com/lidofinance/gate-seals/pull/36/commits/a2e3ecdc732f6fe42a63a98acd32e3c8e293bac8)

Alexvozhak replied:
[done](https://github.com/lidofinance/gate-seals/pull/36/commits/ef7020e0b31388c231128a3f81607a5362806c96)

TheDZhon commented on contracts/GateSealFactory.vy:
let's roll back this change

TheDZhon commented on contracts/GateSealFactoryV2.vy:
```diff
@@ -0,0 +1,88 @@
+"""
+@title GateSealFactoryV2
+@author alex.k@lido.fi
+@notice A factory contract for GateSeals
```
```suggestion
@notice A factory contract for GateSeals (V2)
```

TheDZhon commented on contracts/GateSealFactoryV2.vy:
```diff
@@ -0,0 +1,88 @@
+@title GateSealFactoryV2
+@author alex.k@lido.fi
+@notice A factory contract for GateSeals
+@dev This contract is meant to simplify the GateSeal deploy.
```
```suggestion
@dev This contract is meant to simplify the GateSeal V2 deploy.
```

TheDZhon commented on contracts/GateSealFactoryV2.vy:
```diff
@@ -0,0 +1,88 @@
+"""
+
+event GateSealCreated:
+    gate_seal: address
```
```suggestion
    gate_seal: indexed(address)
```

TheDZhon commented on contracts/GateSealFactoryV2.vy:
```diff
@@ -0,0 +1,88 @@
+# GateSeals were originally designed to pause WithdrawalQueue and ValidatorExitBus,
+# however, there is a non-zero chance that there might be more in the future, which
+# is why we've opted to use a dynamic-size array.
+MAX_SEALABLES: constant(uint256) = 8
```
```suggestion
MAX_SEALABLES: constant(uint256) = 10
```

TheDZhon commented on contracts/GateSealFactoryV2.vy:
```diff
@@ -0,0 +1,88 @@
+    _prolongation_limit: uint256,
+    _prolongation_period_seconds: uint256,
+    _prolongation_window_seconds: uint256,
+    _dao_ops_reserve_seconds: uint256,
```
```suggestion
    _pre_expiration_offset: uint256,
```

TheDZhon commented on contracts/GateSealFactoryV2.vy:
```diff
@@ -0,0 +1,88 @@
+    @param _prolongation_limit number of available prolongations
+    @param _prolongation_period_seconds prolongation period in seconds
+    @param _prolongation_window_seconds prolongation window in seconds
+    @param _dao_ops_reserve_seconds DAO Ops reserve in seconds
```
```suggestion
    @param _pre_expiration_offset prolongation window end offset before the expiration
```

TheDZhon commented on contracts/GateSealV2.vy:
```diff
@@ -0,0 +1,357 @@
+@title GateSealV2
+@author alex.k@lido.fi
+@notice A one-time panic button for pausable contracts
+@dev GateSeal is a one-time immediate emergency pause for pausable contracts.
```
```suggestion
@dev GateSealV2 is a one-time immediate emergency pause for pausable contracts.
```

TheDZhon commented on contracts/GateSealV2.vy:
```diff
@@ -0,0 +1,357 @@
+     GateSeals serve as a long-term safety mechanism. 
+     A committee ensures the GateSeal remains viable by prolonging its duration periodically.
+     Should they fail to do so, the GateSeal will automatically expire, requiring the deployment of a new one.
+     This approach maintains the multisig's limitations in terms of power and duration, while allowing the DAO to bypass the need for an annual full vote.
```
```suggestion
     This approach maintains the multisig's limitations in terms of power and duration, while streamlining the full DAO vote cadence.
```

TheDZhon commented on contracts/GateSealV2.vy:
```diff
@@ -0,0 +1,357 @@
+event Sealed:
+    sealed_by: address
+    sealed_for: uint256
+    sealable: address
```
```suggestion
    sealable: indexed(address)
```

TheDZhon commented on contracts/GateSealV2.vy:
```diff
@@ -0,0 +1,357 @@
+event Prolonged:
+    prolonged_by: address
+    prolongations_remaining: uint256
+    new_expiry: uint256
```
```suggestion
    new_expiry: indexed(uint256)
```

TheDZhon commented on contracts/GateSealV2.vy:
```diff
@@ -0,0 +1,357 @@
+PROLONGATION_WINDOW_SECONDS: immutable(uint256)
+
+# DAO_OPS_RESERVE ensures the DAO can deploy a new GateSeal if prolongation fails
+DAO_OPS_RESERVE_SECONDS: immutable(uint256)
```
PRE_EXPIRATION_OFFSET

TheDZhon commented on contracts/GateSealV2.vy:
```diff
@@ -0,0 +1,357 @@
+    
+        sealable_index += 1
+
+    assert len(failed_indexes) == 0, self._to_error_string(failed_indexes)
```
shall we revert or emit event and succeed though?

Alexvozhak replied:
added seal_all() and seal_some() methods, [here](https://github.com/lidofinance/gate-seals/pull/36/commits/0f0a8a50f73849520b07917dc42d419bf0a6c496)

TheDZhon reviewed (commented):
:eyes:

Alexvozhak committed:
Apply suggestions from code review

Co-authored-by: Eugene Mamin <TheDZhon@gmail.com>

Alexvozhak committed:
chore: rename DAO_OPS_RESERVE_SECONDS to PRE_EXPIRATION_OFFSET

Alexvozhak committed:
feat: increase maximum number of sealables from 8 to 10

Alexvozhak committed:
feat: add dual sealing methods seal_all() and seal_some()
 - add seal_all() for all contracts and seal_some() for partial sealing.
 - extract shared logic to _seal_implementation().
 - update tests accordingly.

Alexvozhak committed:
feat: remove sealable interface validation at deployment time

Alexvozhak committed:
fix: add sequential deployment in generate_sealables to prevent nonce conflicts

Alexvozhak committed:
docs: readme aligned ADR
- add 5-year lifetime limit
- remove duplicate paragraph

Alexvozhak committed:
fix: unify prolongation window logic to expired logic

failingtwice commented on contracts/GateSealFactoryV2.vy:
```diff
@@ -0,0 +1,88 @@
+# @version 0.4.2
+
+"""
+@title GateSealFactoryV2
```
Why do we add new files for factory and gateseal v2? Why not update the old ones?

Alexvozhak replied:
https://github.com/lidofinance/gate-seals/pull/36#discussion_r2136448535

failingtwice commented on contracts/GateSealFactoryV2.vy:
```diff
@@ -0,0 +1,88 @@
+# The maximum number of sealables is 10.
+# GateSeals were originally designed to pause WithdrawalQueue and ValidatorExitBus,
+# however, there is a non-zero chance that there might be more in the future, which
+# is why we've opted to use a dynamic-size array.
```
There are already other contracts in DualGovernance. May wanna cut this comment altogether

Alexvozhak replied:
[done](https://github.com/lidofinance/gate-seals/pull/36/commits/7bc2d6b0f8a7e1d40bb434d5a7759685dd500f2d)

failingtwice commented on contracts/GateSealFactoryV2.vy:
```diff
@@ -0,0 +1,88 @@
+# GateSeals were originally designed to pause WithdrawalQueue and ValidatorExitBus,
+# however, there is a non-zero chance that there might be more in the future, which
+# is why we've opted to use a dynamic-size array.
+MAX_SEALABLES: constant(uint256) = 10
```
e.g. Explanation: Vyper doesn't support dynamic-sized arrays so we settled on this number.

Alexvozhak replied:
[done](https://github.com/lidofinance/gate-seals/pull/36/commits/7bc2d6b0f8a7e1d40bb434d5a7759685dd500f2d)

failingtwice commented on contracts/GateSealFactoryV2.vy:
```diff
@@ -0,0 +1,88 @@
+):
+    """
+    @notice Create a new GateSeal.
+    @dev    All of the security checks are done inside the GateSeal constructor.
```
This ensures that a gateseal will be properly parametrized without factory deploy.

Alexvozhak replied:
[done](https://github.com/lidofinance/gate-seals/pull/36/commits/7bc2d6b0f8a7e1d40bb434d5a7759685dd500f2d)

failingtwice commented on contracts/GateSealV2.vy:
```diff
@@ -0,0 +1,359 @@
+# === SEALABLE CONTRACTS LIMITS ===
+# The maximum number of sealables is 10.
+# GateSeals were originally designed to pause WithdrawalQueue and ValidatorExitBus,
+# however, there is a non-zero chance that there might be more in the future.
```
Not relevant

Alexvozhak replied:
[done](https://github.com/lidofinance/gate-seals/pull/36/commits/7bc2d6b0f8a7e1d40bb434d5a7759685dd500f2d)

failingtwice commented on contracts/GateSealV2.vy:
```diff
@@ -0,0 +1,359 @@
+PROLONGATION_WINDOW_SECONDS: immutable(uint256)
+
+# PRE_EXPIRATION_OFFSET ensures the DAO can deploy a new GateSeal if prolongation fails
+PRE_EXPIRATION_OFFSET: immutable(uint256)
```
GRACE_PERIOD ?

Alexvozhak replied:
https://github.com/lidofinance/gate-seals/pull/36#discussion_r2278697306

Alexvozhak replied:
[done, EXPIRATION_BUFFER_SECONDS](https://github.com/lidofinance/gate-seals/pull/36/commits/6741d6ba0cf87b437b3c55faf306e0782200b6d6)

failingtwice commented on contracts/GateSealV2.vy:
```diff
@@ -0,0 +1,359 @@
+
+# Timeline: Now → [PROLONGATION_WINDOW] → [PRE_EXPIRATION_OFFSET] → Expiry
+# PROLONGATION_WINDOW is the period during which the committee can prolong
+PROLONGATION_WINDOW_SECONDS: immutable(uint256)
```
These two variables sound like they mean exactly the same

Alexvozhak replied:
[done](https://github.com/lidofinance/gate-seals/pull/36/commits/9ce6699e9c10ed775c695e0b8293834e2fc13b60)

failingtwice commented on contracts/GateSealV2.vy:
```diff
@@ -0,0 +1,359 @@
+
+# Total lifetime across all prolongations is capped at five years
+TOTAL_LIFETIME_DAYS: constant(uint256) = 365 * 5
+TOTAL_LIFETIME_SECONDS: constant(uint256) = SECONDS_PER_DAY * TOTAL_LIFETIME_DAYS
```
MAX_LIFETIME ?

Alexvozhak replied:
[done](https://github.com/lidofinance/gate-seals/pull/36/commits/a29521c2d0fc3d72d780b3bc9757368fa8675b10)

failingtwice commented on contracts/GateSealV2.vy:
```diff
@@ -0,0 +1,359 @@
+    self._seal_implementation(_sealables)
+
+@internal
+def _seal_implementation(_sealables: DynArray[address, MAX_SEALABLES]):
```
_seal_implementation might be confusing. Perhaps simply _seal?

Alexvozhak replied:
[done](https://github.com/lidofinance/gate-seals/pull/36/commits/487f460c8aa1562a5b6748bbd99c656d84785d66)

Alexvozhak committed:
docs: update MAX_SEALABLES comments to reflect current implementation

Alexvozhak committed:
chore: rename _seal_implementation to _seal for clarity

Alexvozhak committed:
chore: rename PROLONGATION_PERIOD_SECONDS to PROLONGATION_EXTENSION_SECONDS

Alexvozhak committed:
chore: rename TOTAL_LIFETIME to MAX_LIFETIME

Alexvozhak committed:
chore: rename PRE_EXPIRATION_OFFSET to EXPIRATION_BUFFER_SECONDS

Alexvozhak committed:
feat(tests): add comprehensive test coverage

- Add test for view functions accessibility after sealing/expiry
- Add test for multiple prolongations with limit > 1
- Add test for prolongation window reopening logic
- Add test for max lifetime architectural constraint

tamtamchik commented:
Closed as GateSeal V2 deprecated.
